dns/resolver.py

```diff
@@ -58,27 +58,50 @@ class YXDOMAIN(dns.exception.DNSException):
 Timeout = dns.exception.Timeout
 
 class NoAnswer(dns.exception.DNSException):
-    """The response did not contain an answer to the question."""
-    pass
+    """The DNS response does not contain an answer to the question."""
+    def __init__(self, question=None):
+        super(dns.exception.DNSException, self).__init__()
+        self.question = question
+
+    def __str__(self):
+        message = self.__doc__
+        if self.question:
+            message = message[0:-1]
+            for q in self.question:
+                message += ' %s' % q
+        return message
+
 
 class NoNameservers(dns.exception.DNSException):
     """No non-broken nameservers are available to answer the query."""
-    pass
+    def __init__(self, errors=[]):
+        """Optionally construct message with list of servers and errors.
+
+        @param errors: list of servers and respective errors
+        @type errors: [(server ip address, any object convertible to string)]
+        """
+        super(dns.exception.DNSException, self).__init__()
+        self.errors = errors
+
+    def __str__(self):
+        message = self.__doc__
```

Comment:
the __doc__ is maybe a bit much in this case as it also contains the @param stuff.